File: cpp/src/advanced/mro.cpp
#include "mro.h"



void A::method() {cout << "Metodo di A L0" << endl; }

void B::method() { 
  cout << "Metodo di B L1" << endl;
  A::method();  // chiamata diretta ad A
}

void C::method()  { 
  cout << "Metodo di C L1" << endl;
  A::method();  // chiamata diretta ad A
}

void D::method()  {
  cout << "Metodo di D L2" << endl;
  B::method();  // manual resolution order
  C::method(); 
}

void mro_example() { D d; d.method(); }

File: Python/src/advanced/mro.py
# Method Resolution Order
class A:
  def method(self):
    print("Metodo di A L0")
class B(A):
  def method(self):
    print("Metodo di B L1")
    super().method() 
class C(A):
  def method(self):
    print("Metodo di C L1")
    super().method() 
class D(B, C):
  def method(self):
    print("Metodo di D L2")
    B.method(self)  # evito ambiguità con chiamata diretta
    C.method(self) 
    print("Metodo chiamata super(), precedenza ereditarietà sinistra") 
    super().method() 

class TestMRO:
  @classmethod
  def run(cls):
    d = D()
    d.method()